### ISSUE TITLE
Warmups are skipped when running a benchmark suite

### ISSUE THREAD

**Issue by willow-ahrens** [JuliaCI/BenchmarkTools.jl#291]
If we tune the suite, then everything is warmed up all at once, but we don't get warmups before measuring individual benchmarks.
If we use cached parameters, then all the benchmarks are cold when we start the suite.

Ideally, we should warmup before running each individual benchmark in a suite.

**Comment by willow-ahrens** [comment]
Sure thing!

TL;DR: The first execution of code runs slower than the subsequent executions of code, and running other code in between changes the runtime. `@benchmark` and friends all run a warmup execution to avoid measuring the first run, but if you have a suite of benchmarks for which you specify parameters, calling `run` on the suite would not warm up any of those benchmarks. Calling `warmup` on the suite would warm all of them up separately first, which is different than warming them up individually. Also, it is not made clear that it is best practice to call `warmup` on the suite. A suite of benchmarks functions differently than many individual `@benchmark` statements, and the current interface provides no way for them to be equivalent.

A few terms that I've seen used around here:

- A kernel is the code we want to benchmark
- An execution is a single run of that code
- A sample is the time it takes to run a fixed number of executions in a loop

`@btime`, `@benchmark`, and `@belapsed` all benchmark a single kernel eagerly. [Here](https://github.com/JuliaCI/BenchmarkTools.jl/blob/be72665530ef6328d39e70ffa91e286d14f84c19/src/execution.jl#L388), they start by running a warmup, then tuning the number of executions per sample, then they run the benchmark a maximum of 5 seconds or 10000 executions, whichever happens first.

The purpose of the warmup is to make sure that the tuning and measurements don't include that first run of the benchmark. If we imagine the kernel takes 20s to compile (maybe it loads a big package or has some other cached startup cost), then if we measure without warming up or tuning, the first sample will take 20 seconds (exceeding 5 s) and the returned runtime will be 20s, even if subsequent samples would have been much faster (e.g. 0.1 s).

The warmup also loads things into cache. If we are benchmarking a linear algebra kernel like matrix-vector multiply and the matrix fits into cache, then the first execution would load the matrix into cache (very slow), but subsequent executions would already have the matrix already loaded and run much faster.

Currently, if you follow the docs and construct a benchmark suite, then call `tune!` and `run`, `tune!` would measure the first ever execution (with compilation), and `run` would measure a cold execution (not in cache).

If you call `tune!` and then `loadparams!`, then `run` would measure the first ever execution (with compilation)

If you call `tune!` and then `loadparams!`, then `warmup` then `run`, `run` would measure a cold execution (not in cache).

There is no way to tell BenchmarkTools that we should `warmup`, then `tune!`, then `run` each benchmark individually. Also, I'd argue that the warming up should really be done automatically, because it is done automatically in `@benchmark` and users may not realize they need to do it if they are using a suite.

**Comment by gdalle** [comment]
Thank you for the detailed explanation! Not sure anything will be done about it soon but at least it's clearer to me now, and hopefully to anyone willing to contribute a fix

**Comment by willow-ahrens** [comment]
I might be interested in contributing a fix sometime. Is there community consensus that we want this fixed? Would there be interest in reviewing such a PR?

**Comment by gdalle** [comment]
As you can tell by the number of reactions on your issue, there isn't much of a consensus, nor much of a community to begin with :rofl: I became a maintainer this weekend without asking for it, just because I contributed a minor docs fix and @vchuravy nerdsniped me ^^

So definitely feel free to contribute, I will review to the best of my ability, but since I know next to nothing about macros, there will be a learning curve. If people want to weigh in they will, if not we'll just debate this between ourselves

**Comment by willow-ahrens** [comment]
The quickest fix would be 
1. adjust the `run` and `lineartrial` functions to call `samplefunc` once as a warmup before collecting info.
2. Deprecate warmup and remove any explicit calls to warmup.
3. Docs and tests

**Comment by willow-ahrens** [comment]
TL;DR: My proposal is to just make warming up a part of any function that benchmarks code.

**Comment by gdalle** [comment]
I think it's a good idea to make warmup the default, since there is no situation I can think of where we want to include the compile time (otherwise we would just use `@time`). What do other people think?

**Comment by willow-ahrens** [comment]
@gdalle is two weeks enough to conclude there are no objections?

**Comment by willow-ahrens** [comment]
@gdalle I think it's safe then to start on a pull request, do you agree?

**Comment by gdalle** [comment]
Yeah knock yourself out 😉

**Comment by gdalle** [comment]
Sorry for the delay. And once we get it merged I'll do a bit of cleanup and tag v2.0

**Comment by Zentrik** [comment]
I've noticed that it seems the first eval when benchmarking with BaseBenchmarks is significantly slower than subsequent evals, I assumed it was compilation time. Looking at BaseBenchmarks they seem to do `loadparams` but don't have a warmup so this should fix that.

**Comment by willow-ahrens** [comment]
Thanks @gdalle!
